PM-3231 Vault Onboarding Part 1

Jingo88 opened the pull request:
```
- [ ] Bug fix
- [ X ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Add the onboarding component to the individual vault page so new users can be directed to import data, create a new item, and install the browser extension.

* create account is always true
* import data will link to import page
* add new item link will open new item modal
* import data section is dynamic based on users individual vault policy
* install new extension link will navigate to browser specific extension install
* dismiss button will close onboarding and will keep it closed on refresh

NOTE: The bullet inside testing notes stating `Show onboarding UI as complete when all items are completed until the user logs into a new session` cannot be shown yet. The `install extension` completion logic will take place in M2 of this onboarding feature. [Ticket Here](https://bitwarden.atlassian.net/browse/PM-4109)

## Code changes

* Moved onboarding files to web shared folder
* Removed `sm-` prefix on onboarding selectors
* Created `vault-onboarding` component inside individual vault. This wraps the reusable onboarding in the shared folder
* Spec file created for new component
* The `vault.component` will show the `vault-onboarding` component based on the `FeatureFlag.VaultOnboarding` feature flag as well as the `showOnboarding` logic
* New strings added to `messages.json`
* New get and set vault onboarding tasks added to `state.service`
* Import data section of vault onboarding is dynamic based on individual vault policy status


## Video

https://github.com/bitwarden/clients/assets/8302660/1b420840-1c60-42ec-a98c-f90029a97842

LRNcardozoWDF reviewed (approved):
LGTM!

shane-melton reviewed (changes requested):
This is looking really good, nice work! Only one issue with the current implementation around the individual vault policy and accepted, but not confirmed, users.

Otherwise, I just had some suggestions/nits to potentially make things a bit cleaner.

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -0,0 +1,180 @@
+  }
+
+  ngOnInit() {
+    this.checkCreationDate();
```
issue: I'm concerned that we're adding an API call every time this component is rendered on the vault to get the account's age. Can we adjust the logic to save the onboarding tasks as completed for "old" accounts?

That would then allow us to check if the tasks have been marked as completed **before** attempting to call the API to get the account's profile/age. (e.g. if the tasks are marked as completed then we don't care about the account's age and can skip the API call).

shane-melton commented on apps/web/src/app/vault/individual-vault/vault.component.ts:
```diff
@@ -149,6 +149,10 @@ export class VaultComponent implements OnInit, OnDestroy {
     FeatureFlag.BulkCollectionAccess,
     false,
   );
+  protected showOnboardingAccess$ = this.configService.getFeatureFlag$(
```
question/suggestion: Can't we also move this into the vault-onboarding component? I'm not seeing any reason to grab it at the vault level and pass it in as a attribute.

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -0,0 +1,180 @@
+    if (changes?.ciphers) {
+      const { currentValue, previousValue } = changes.ciphers;
+      if (this.showOnboarding && currentValue?.length !== previousValue?.length) {
+        this.saveCompletedTasks({
```
question: This looks like it will "uncomplete" a task whenever the `ciphers` list is empty even if temporarily. Is this the desired behavior? If so, we may want to look into how this behaves when a sync fails or `ciphers` is empty initially. 

Currently, it seems like this will set the `importData` task to false and then show the onboarding component, then whenever the sync completes it update the task again and hide the onboarding component.

shane-melton replied:
Also, if we mark tasks as complete for "old" accounts (see my other comment above in `ngOnInit`) we'll want to be careful with how we toggle these tasks for those old accounts. Otherwise, we may still end up making the unnecessary API calls.

Jingo88 replied:
The import data task is tied to if ciphers exist. if a cipher is created and then deleted leaving the vault empty the import data task will revert to "incomplete". this is desired behavior. 

If the onboarding component values are all true (fully completed, dismissed, an old acct) than it's values such as `importData` will never be touched and it won't show again. 

I've updated the `checkCreationDate` method to only be called if `showOnboarding === true` so that should save on unnecessary API calls to `getProfile` 

I also am now calling the `hideOnboarding()` method to set local storage once we determine if an account is `old` so after their initial login their `showOnboarding` status should always be false and nothing would run. 

Adding a new screen recording to the description for clarity

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -0,0 +1,180 @@
+
+  navigateToImport() {
+    if (!this.isIndividualPolicyVault) {
+      this.onboardingTasks$.pipe(take(1), takeUntil(this.destroy$)).subscribe((onboardingTasks) => {
```
suggestion:  We can avoid the subscription here and just `await firstValueFrom(this.onboardingTasks$)`.

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -0,0 +1,180 @@
+  navigateToImport() {
+    if (!this.isIndividualPolicyVault) {
+      this.onboardingTasks$.pipe(take(1), takeUntil(this.destroy$)).subscribe((onboardingTasks) => {
+        if (!onboardingTasks.importData) {
```
question: Why do we not navigate to import if the task is completed? Won't it still appear as a clickable link on the page?

Jingo88 replied:
thanks. removing this logic. the import data link will still be clickable and navigate to tools/import

shane-melton reviewed (changes requested):
Nice work on this! I love the addition of tests! 

I had a few more comments and one concern with hitting the account API endpoint on every page load. Let me know if you have any questions!

Thomas-Avery reviewed (commented):
This would change code ownership of the onboarding components from the SM team to the platform team.

Will want to confirm they're good with taking over ownership.

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.html:
```diff
@@ -16,7 +16,9 @@
       <p class="tw-pl-1">
-        {{ "onboardingImportDataDetails" | i18n }}
+        {{ "onboardingImportDataDetailsPartOne" | i18n }}
+        <a (click)="emitToAddCipher()" class="link"> new item </a>
```
issue: "new item" needs to be translated

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -35,18 +37,15 @@ export type VaultOnboardingTasks = {
   private destroy$ = new Subject<void>();
   isNewAccount: boolean;
-  private readonly onboardingReleaseDate = new Date(2023, 12, 22);
+  private readonly onboardingReleaseDate = new Date("2023-12-22");
```
issue: We need to update this to the release we expect this to go out.

Jingo88 replied:
Yep. Will keep as is for QA. Planning to update the date after M2 is code complete and QA Approved.

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -153,24 +161,20 @@ export class VaultOnboardingComponent implements OnInit, OnChanges, OnDestroy {
-          this.router.navigate(["tools/import"]);
-        }
-      });
+      this.router.navigate(["tools/import"]);
```
question: Should we `await` this? I feel like I always see warnings about un-awaited promises with regards to the router.

Jingo88 replied:
I think it's okay without an `await`, since there isn't any other code that needs to run before the navigation.

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -58,19 +57,16 @@ export class VaultOnboardingComponent implements OnInit, OnChanges, OnDestroy {
 
   ngOnInit() {
-    this.checkCreationDate();
+    this.checkOnboardingFlag();
```
question: Should we `await` these `async` method calls? I'm worried we may run into race conditions if they depend on being run in a particular order?

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -130,6 +134,10 @@ export class VaultOnboardingComponent implements OnInit, OnChanges, OnDestroy {
     }
+
+    if (this.showOnboarding) {
+      this.checkCreationDate();
```
question: Similarly, should we `await` this? Otherwise I'm worried the onboarding component is going to flash on the screen briefly while it's checking the account profile date.

shane-melton reviewed (changes requested):
Looking very close, nice work! Only a couple questions about the `async` methods not being `awaited`  and a missing translation.

shane-melton reviewed (approved):
Looks good to go! Thanks for making those changes, nice work!

willmartian commented on apps/web/src/app/shared/components/onboarding/onboarding.stories.ts:
This story needs to be updated to use the updated selector prefix `app-`

willmartian commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.html:
```diff
@@ -0,0 +1,37 @@
+<div *ngIf="isNewAccount && showOnboarding && showOnboardingAccess" class="tw-mb-6">
+  <app-onboarding
+    *ngIf="onboardingTasks$ | async"
```
Rather than reapplying the `async` pipe below, you could do `onboardingTasks$ | async as onboardingTasks`

willmartian commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.html:
```diff
@@ -0,0 +1,37 @@
+    >
+      <p class="tw-pl-1">
+        {{ "onboardingImportDataDetailsPartOne" | i18n }}
+        <a (click)="emitToAddCipher()" class="link">
+          {{ "onboardingImportDataDetailsLink" | i18n }}
+        </a>
```
Avoid mixing Tailwind and Bootstrap in new components (`class="link"`). 

Also avoid using anchor elements unless we are navigating to a different route.

We can use `<button bitLink>` here: https://components.bitwarden.com/?path=/docs/component-library-link--docs

willmartian commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.html:
```diff
@@ -1,25 +1,31 @@
+  class="tw-mb-6"
+>
   <app-onboarding
     *ngIf="onboardingTasks$ | async"
```
```suggestion
    *ngIf="onboardingTasks"
```

willmartian reviewed (approved):
Looks good to me! One last non-blocking nit pick:

willmartian reviewed (changes requested):
Hey @Jingo88, I apologize for the change in review, but @justindbaur pointed out to me that we should generally be using the new `StateProvider` approach instead of adding fields to the `StateService` monolith. 

- See docs [here](https://add-state-provider-framework.contributing-docs.pages.dev/architecture/deep-dives/state/)
- [Example PR](https://github.com/bitwarden/clients/pull/7543) using the new paradigm

Let @bitwarden/team-platform-dev know if you have any questions!

Jingo88 commented:
Updated the PR to use the new state provider framework. Updated the video in the description. All onboarding features working as expected.

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/services/vault-onboarding.service.ts:
```diff
@@ -0,0 +1,38 @@
+  installExtension: boolean;
+};
+
+const VAULT_ONBOARDING_KEY = new KeyDefinition<any>(VAULT_ONBOARDING, "data", {
```
suggestion (non-blocking): How would you feel about giving this key a slightly more descriptive name like `"tasks"` instead of `"data"`?

shane-melton replied:
suggestion: We should avoid using `any` for the KeyDefinition. I think we should be able to use `VaultOnboardingTasks` instead.

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/services/vault-onboarding.service.ts:
```diff
@@ -0,0 +1,38 @@
+
+@Injectable()
+export class VaultOnboardingService implements VaultOnboardingServiceAbstraction {
+  private vaultOnboardingState: ActiveUserState<Record<string, VaultOnboardingTasks>>;
```
question: Why is this a `ActiveUserState<Record<string, VaultOnboardingTasks>>` and not `ActiveUserState<VaultOnboardingTasks>`? If we remove the `Record<>` it would let us avoid a lot of the de-structuring when accessing the state in this service and the component.

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -48,31 +44,32 @@ export class VaultOnboardingComponent implements OnInit, OnChanges, OnDestroy {
+    this.onboardingTasks$ = this.vaultOnboardingService.vaultOnboardingState$;
+    this.onboardingTasks$
+      .pipe(takeUntil(this.destroy$))
+      .subscribe((tasks: { vaultTasks: VaultOnboardingTasks }) => {
```
suggestion: If we remove the `Record<>` from `this.vaultOnboardingService.vaultOnboardingState$` we can avoid the anonymous type definition here in the subscribe method.

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -48,31 +44,32 @@ export class VaultOnboardingComponent implements OnInit, OnChanges, OnDestroy {
+        this.showOnboarding =
+          tasks !== null ? Object.values(tasks.vaultTasks).includes(false) : true;
+        this.currentTasks = tasks !== null ? tasks.vaultTasks : null;
+        this.setOnboardingTasks();
```
question: We're setting the onboarding tasks in a subscription to the onboarding tasks observable, any concern for an infinite loop here?

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -119,33 +116,26 @@ export class VaultOnboardingComponent implements OnInit, OnChanges, OnDestroy {
-    this.stateService.setVaultOnboardingTasks({
-      currentStatus: vaultTasks,
-    });
+    this.vaultOnboardingService.setVaultOnboardingTasks(vaultTasks);
```
suggestion: Missing `await` here.

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/vault-onboarding.component.ts:
```diff
@@ -119,33 +116,26 @@ export class VaultOnboardingComponent implements OnInit, OnChanges, OnDestroy {
-  async setOnboardingTasks() {
-    const tasksInStorage: any = (await this.stateService.getVaultOnboardingTasks()) || null;
-    if (tasksInStorage == null) {
+  setOnboardingTasks() {
```
question: Why did we remove `async` from this method? It still has some async actions within it that are no longer being awaited.

shane-melton commented on libs/common/src/enums/feature-flag.enum.ts:
```diff
@@ -5,6 +5,7 @@ export enum FeatureFlag {
   FlexibleCollectionsV1 = "flexible-collections-v-1", // v-1 is intentional
   BulkCollectionAccess = "bulk-collection-access",
   VaultOnboarding = "vault-onboarding",
+  GeneratorToolsModernization = "generator-tools-modernization",
```
thought: This seems unrelated to the onboarding component, we might want to double check that we didn't overwrite something from `main`.

shane-melton reviewed (changes requested):
I think we're almost there, nice work! I have a few comments/question on they types we're using with the state provider changes.

willmartian commented on apps/web/src/app/shared/components/onboarding/onboarding.module.ts:
```diff
@@ -0,0 +1,23 @@
+  imports: [SharedModule, ProgressModule],
+  exports: [OnboardingComponent, OnboardingTaskComponent],
+  declarations: [OnboardingComponent, OnboardingTaskComponent],
+  providers: [
+    {
+      provide: VaultOnboardingServiceAbstraction,
+      useClass: VaultOnboardingService,
+    },
+  ],
```
The shared onboarding components are 1.) purely presentational and 2.) used outside of the Vault team. As such, I don't think we should add references to a Vault-specific service to it. 

Perhaps this should be added to the `VaultModule` or `VaultOnboardingComponent` instead?

willmartian reviewed (dismissed):
Platform changes LGTM! Thank you @Jingo88

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/services/vault-onboarding.service.ts:
```diff
@@ -16,23 +16,23 @@ export type VaultOnboardingTasks = {
+  async setVaultOnboardingTasks(newState: VaultOnboardingTasks): Promise<void> {
     await this.vaultOnboardingState.update(() => {
-      return { vaultTasks: { ...newState } };
+      return { ...newState };
```
question: Any particular reason for the `...` spread operation here?

shane-melton commented on apps/web/src/app/vault/individual-vault/vault-onboarding/services/vault-onboarding.service.ts:
```diff
@@ -16,23 +16,23 @@ export type VaultOnboardingTasks = {
   constructor(private stateProvider: StateProvider) {
     this.vaultOnboardingState = this.stateProvider.getActive(VAULT_ONBOARDING_KEY);
     this.vaultOnboardingState$ = this.vaultOnboardingState.state$;
   }
```
nit (non-blocking): We can move the assignment of this state and observable to its declaration and out of the constructor to keep things together and is how we've done the other state provider migrations in Vault.

```suggestion
  private vaultOnboardingState = this.stateProvider.getActive(VAULT_ONBOARDING_KEY);
  vaultOnboardingState$ = this.vaultOnboardingState.state$;
  
  constructor(private stateProvider: StateProvider) { }
```

Jingo88 replied:
Thanks. I'll note this and make the change in the follow-up onboarding M2 work

shane-melton reviewed (approved):
Looks good, nice work! Just a small non-blocking question and nit pick, neither of which should hold this up from QA.

Jingo88 commented:
Small Updates to address A11y and Browser Extension links for the QA findings below

* added a optional input of "isDisabled" to onboarding component so we can make a task unclickable
* Updated the navigation to `tools/import` so the task is rendered with an `<a>`
* Added other compatible browser links to the install task

https://bitwarden.atlassian.net/browse/PM-6118
https://bitwarden.atlassian.net/browse/PM-6119
https://bitwarden.atlassian.net/browse/PM-6120

shane-melton reviewed (approved):
Looks good to go! 🎉 🚀
